### PR TITLE
Plugins: Add Command Palette extension point

### PR DESCRIPTION
**Pull request by sd2k**
**What is this feature?**

It would be cool for plugins to be able to add items to the command palette. This PR adds a POC of that functionality. I've created a draft PR at https://github.com/grafana/machine-learning/pull/3501 which uses it and it seems to work well.

**Why do we need this feature?**

This makes plugin functionality easier to find from throughout Grafana.

**Who is this feature for?**

Plugin developers and users of plugins.

**Which issue(s) does this PR fix?**:

N/A

**Special notes for your reviewer:**

I'm not sure what the priority of extension links should be inside the command palette, probably not '6' though since they all appear right at the top.

The context passed to the extension point is just an empty object for now because I don't think we can really pass much in; the palette could be opened from any page and may not even have a time range selected. Is there actually anything useful we can guarantee will be there? 🤷 

---

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by joshhunt** [comment]
Neat!

I think our preferred approach would be to add an extension point to add to the Create New menu, and then make sure that shows up in command pallette 

image

File: packages/grafana-data/src/types/pluginExtensions.ts [hunk 1]
```diff
@@ -117,6 +117,7 @@ export type PluginExtensionEventHelpers<Context extends object = object> = {
 
 // Extension Points available in core Grafana
 export enum PluginExtensionPoints {
+  CommandPalette = 'grafana/commandpalette',
```
Comment (mckn) [inline packages/grafana-data/src/types/pluginExtensions.ts:120]:
```suggestion
  CommandPalette = 'grafana/commandpalette/action',
```

Comment (mckn) [reply packages/grafana-data/src/types/pluginExtensions.ts:120]: Just in case we want to add extensions to other parts of the command palette.

**Review by mckn — COMMENTED** [review]
Added one small nit but overall I think this looks great!

File: public/app/features/commandPalette/actions/extensionActions.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import { CommandPaletteAction } from '../types';
+import { EXTENSIONS_PRIORITY } from '../values';
+
+export default function getExtensionActions(): CommandPaletteAction[] {
+  const context: PluginExtensionCommandPaletteContext = {};
+  const { extensions } = getPluginLinkExtensions({
+    extensionPointId: PluginExtensionPoints.CommandPalette,
+    context,
+    limitPerPlugin: 3,
```
Comment (leventebalogh) [inline public/app/features/commandPalette/actions/extensionActions.ts:12]: I think it's good for now, we might want to revisit this in the future (maybe we can allow a higher limit for plugins here as this list can be searched).

**Review by leventebalogh — APPROVED** [review]
LGTM, nice one! 🚀

**Comment by leventebalogh** [comment]
> I think our preferred approach would be to add an extension point to add to the Create New menu, and then make sure that shows up in command pallette

I kinda disagree on this @joshhunt, at least in the current form of the "Create new" menu. While the UX of the command palette wouldn't "suffer" if plugins start introducing commands to it (it's searchable with an actively focused input when you open it), the "Create new" menu would quickly become unusable if 10 new actions appear in it. I would rather separate the two extension points.

Probably also something to bring up on the next UX feedback sessions?

File: public/app/features/commandPalette/values.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+export const EXTENSIONS_PRIORITY = 6;
```
Comment (sd2k) [inline public/app/features/commandPalette/values.ts:1]: It's worth considering what this priority should be; this puts extensions right at the top which is probably a bit extreme, maybe somewhere in the middle makes sense?

**Comment by leventebalogh** [comment]
> @leventebalogh It's been Frontend Platform's goal, guided by UX and Product, since the beginning to allow plugins to add their own actions to the 'Add new' menu. This has been on our backlog (metaphorically, no issue) since before grafana's even had extension points.

I see, I didn't know that, thanks for the info.

> I'll discuss this with team today.

Thanks!

> In general we're supportive of extension points for plugins to add to Command Palette, but for this in particular we would like to add it to the 'Add new' menu instead, and so would like to do that first and when the need arrises to add to Command Palette specifically we can do that.

Taking that into account, I guess we have the following possible options:

1. **We only make the "Add new" menu extendable for now**
    Extensions registered to the "Add new" menu could also automatically be available in the command palette, too.
    
2. **We create two extension points (one for the "Add new" and one for the command palette)**
    Either in this PR or by creating a separate one for the "Add new" extension point. The ML app could register its link extension to both extension points, so they appear in both places. This would also mean that we can set stricter "limitations" on the "Add new" menu than on the command palette.
    
I would personally vote for the second one, but let's see what the team thinks.

**Comment by mckn** [comment]
> @leventebalogh It's been Frontend Platform's goal, guided by UX and Product, since the beginning to allow plugins to add their own actions to the 'Add new' menu. This has been on our backlog (metaphorically, no issue) since before grafana's even had extension points.
> 
> I'll discuss this with team today.
> 
> In general we're supportive of extension points for plugins to add to Command Palette, but for this in particular we would like to add it to the 'Add new' menu instead, and so would like to do that first and when the need arrises to add to Command Palette specifically we can do that.

Ok ok, sounds like the PM's probably should have a chat to align on this. I was under the impression that people from product are driving this functionality by adding Sift support to Grafana. Maybe @sympatheticmoose can help out with this?

Regardless, we can open another PR where we create an extension point for the "Add new" menu if you feel that it should be prioritised before the command palette. It isn't too much work to get that in place.

**Comment by joshhunt** [comment]
> Either in this PR or by creating a separate one for the "Add new" extension point. The ML app could register its link extension to both extension points, so they appear in both places. 

It's the intention that everything from the 'Add new' menu will automatically appear in the Command Palette - that's what currently happens with the current implementation, but adding extension points to Add new might change that up a bit. It's always been the goal that all 'Add new' items are keyboard accessible via command palette.

> Ok ok, sounds like the PM's probably should have a chat to align on this. I was under the impression that people from product are driving this functionality by adding Sift support to Grafana.

It's not clear to me what specific product demands impact this decision. What would be the result of the PMs "aligning" about this? Would we decide to do something different about the command palette/add new?

It's my strong preference to right now add just an extension point for Add new, and have those automatically show up in command palette. If the need arises for an additional extension point to add an item to just the command paletee, we can add that then (we're very supportive of this, but want to see a direct need first!)

@mckn @leventebalogh

**Comment by sd2k** [comment]
Looking into adding this into both places: it looks like both the [command palette](https://github.com/grafana/grafana/blob/f243bd9716c1787fd9d98511a309494c3799d166/public/app/features/commandPalette/actions/useActions.ts#L14) (CP) and [add new](https://github.com/grafana/grafana/blob/f243bd9716c1787fd9d98511a309494c3799d166/public/app/core/components/AppChrome/QuickAdd/QuickAdd.tsx#L19) (AN) button are pulling the actions from the `navBarTree` slice of the Redux store, but I don't think we want to add these extension actions to Redux, right? I'm thinking of adding a function which takes that slice of state, adds the extension actions, and converts them both to something that the CP and AN button can handle. Does that sound like a sensible approach?

**Comment by zsofiakm** [comment]
After speaking with @sympatheticmoose and @joshhunt, we came to these conclusions:
- We'll need extension points in both the command palette and quick-add (add new dropdown)
- Sift investigation is very context-specific to an incident, a panel or a query in Explore, so indeed it makes less sense as a global action in the quick-add. The command palette will allow users to launch it using their keyboard only. So we're on board to add it to the command palette.

@sd2k, will you be adding an extension point only to the command palette, or could you add it to the quick-add as well, so that it can be leveraged by other plugins in the future?

**Comment by sd2k** [comment]
Thanks for the great discussion everyone, and for the summary @zsofiakm.

Just to clarify, am I right in thinking that we'll have two separate extension points, and that everything added to the Quick Add extension point should also appear in the Command Palette extension point, but not vice versa?

Either way, let me know how I should proceed with this PR 🙂

**Comment by sympatheticmoose** [comment]
> am I right in thinking that we'll have two separate extension points

Correct

> everything added to the Quick Add extension point should also appear in the Command Palette extension point

I'm not sure. As I understand, by being added to the Quick Add menu, based on how that works as adding a nav point (pls excuse my potentially incorrect terminology), then it would automatically become available in the command palette. 

So with this model - adding to the command palette extension just does what it says on the tin, adding to the Quick Add extension point also makes it available within the command palette but not explicitly registered to that extension point. Hope this makes sense?

> let me know how I should proceed with this PR

My take would be to scope _this_ PR to the command palette extension point (as original). Given Sift wouldn't be added to the quick add, may make sense for Incident to own the change to add that ext. point as they would likely be the first consumer?

**Comment by sd2k** [comment]
Great, thanks @sympatheticmoose, that makes sense and aligns with my understanding. I think the only outstanding thing on this PR is to decide what priority extensions should get in the command palette then!

**Review by leventebalogh — APPROVED** [review]

**Pull request merged**